src/server/services/wcs/qgswcsgetcapabilities.cpp

```diff
@@ -37,10 +37,40 @@ namespace QgsWcs
   void writeGetCapabilities( QgsServerInterface *serverIface, const QgsProject *project, const QString &version,
                              const QgsServerRequest &request, QgsServerResponse &response )
   {
-    QDomDocument doc = createGetCapabilitiesDocument( serverIface, project, version, request );
+    QgsAccessControl *accessControl = serverIface->accessControls();
+
+    QDomDocument doc;
+    const QDomDocument *capabilitiesDocument = nullptr;
+
+    QgsServerCacheManager *cacheManager = serverIface->cacheManager();
+    if ( cacheManager )
+    {
+      if ( cacheManager->getCachedDocument( &doc, project, request, accessControl ) )
+      {
+        capabilitiesDocument = &doc;
+      }
+    }
+
+    if ( !capabilitiesDocument ) //capabilities xml not in cache. Create a new one
+    {
+      doc = createGetCapabilitiesDocument( serverIface, project, version, request );
+
+      if ( cacheManager )
+      {
+        if ( cacheManager->setCachedDocument( &doc, project, request, accessControl ) &&
+             cacheManager->getCachedDocument( &doc, project, request, accessControl ) )
+        {
+          capabilitiesDocument = &doc;
+        }
+      }
+      if ( !capabilitiesDocument )
+      {
+        capabilitiesDocument = &doc;
+      }
```

Comment:
Why is this needed to re-read the document in the cache? 

It seems more straightforward to have something like:

``` c++
QgsServerCacheManager *cacheManager = serverIface->cacheManager();

// read the document in cache or create it from scratch
if ( cacheManager && cacheManager->getCachedDocument( doc, project, request, accessControl ) ) 
{
  capabilitiesDocument = &doc;
}
else
{
  capabilitiesDocument = createGetCapabilitiesDocument( serverIface, project, version, request );
}

// update the cache
if ( cacheManager )
{
  cacheManager->setCachedDocument( capabilitiesDocument, project, request, accessControl );
}
```

What do you think?

Reply:
It was the way QgsCapabilitiesCache was used.

And I prefer to update the cache only if the capabilities document was not in cache:

```
QgsServerCacheManager *cacheManager = serverIface->cacheManager();

// read the document in cache or create it from scratch
if ( cacheManager && cacheManager->getCachedDocument( doc, project, request, accessControl ) ) 
{
  capabilitiesDocument = &doc;
}
else
{
  capabilitiesDocument = createGetCapabilitiesDocument( serverIface, project, version, request );
// update the cache
  if ( cacheManager )
  {
    cacheManager->setCachedDocument( capabilitiesDocument, project, request, accessControl );
  }
}
```

Reply:
> And I prefer to update the cache only if the capabilities document was not in cache:

Sounds good :+1: